File: ml-benchmark/autolifter/single-pass/max.f
Config ExtraGrammar = "AutoLifter";

Inductive List = cons {Int, List} | nil Unit;

single_pass = \v: List -> Int.
  let run = fix (
    \f: List -> Compress List. \xs: List.
    match xs with
      nil _ -> xs
    | cons {h, t} -> cons {h, f t}
    end
  ) in \xs: List.
  v (run xs);

inf = 100;

/*User provided programs*/

max = fix (\f: List -> Int. \xs: List.
  match xs with
    nil _ -> - 0 inf
  | cons {h, t} ->
    let res = f t in if < res h then h else res
  end
);

main = single_pass max;

File: ml-benchmark/autolifter/single-pass/max.ml
type list_ = Cons of int * list_ | Nil of unit

let single_pass v =
  let rec run xs =
    match xs with
    | Nil _ -> xs
    | Cons (h, t) -> Cons (h, run t)
  in
  fun xs -> v (run xs)

let inf = 100

let rec max xs =
  match xs with
  | Nil _ -> 0 - inf
  | Cons (h, t) ->
    let res = max t in
    if res < h then h else res

let program = single_pass max